### DISCUSSION TITLE
Stretch/Compress Mode Feature

### DISCUSSION THREAD

**Discussion by CopinChats** [audacity/audacity#2599] [Ideas & Feature Requests]
Hello, everyone ! I had an idea that might be helpful for those want to have perfectly synchronized rhythms with different musics. It would be, then, a lot easier : having a Stretch/Compress Mode feature. The goal of this mode is simple : stretch or compress the selected part of the sound to fit perfectly with another sound. The following examples will help you visualize what I'm saying.

Example 1 : you select the zone from the extreme right to an aleatory position, you move your cursor to the right, left click, and drag to the left. The whole selection is compressed to the left.

Example 2 : you repeat the three first actions from the example 1, but this time to drag to the right. The whole selection is stretched to the right.

Example 3 (might be a bit more difficult) : you select the center zone, you move the cursor to the side of the selected zone you want, left click, and drag to the left or right. The non-selected zones are automatically fitting with the selected zone.

There is a basic description of my idea. Feel free to react.

**Comment by rbdannenberg** [comment]
Something like this was implemented in an experimental conditionally-compiled bunch of code for note tracks. There was support for automatic alignment to audio and the ability to tweak the results for fine tuning where necessary. This works well for note tracks which have time maps as part of the data. Audio is different because there is no (current) per-track time deformation curve, so each time you drag, you do significant alterations to the audio, and after multiple drags I think there would be a lot of loss in quality.

Adding time deformation would complicate a lot of things including the GUI, and it's rather specialized, so it sounds unlikely to get a lot of support as a new feature.

A quick-and-dirty approach that might actually be quite useful would be a plug-in that gets data from an editable label track. You put a label track between two tracks and make labels with start and end points corresponding to the current time position of the upper track and the desired position of upper track. I guess you need three labels: ">" means stretch the signal left of the label so that the point at the start of the label is stretched to the endpoint of the label. "<" means shorten the signal left of the label so that the point at the end of the label moves to the beginning of the label. "" (empty) means preserve the left and right times of the label. From these anchor points, you could construct a continuous mapping from old sound to new sound, and Nyquist's phase vocoder, at least for mild stretch and compress could be used and sounds good.  If you didn't like the results, you could undo, adjust a label, and try again, so you would not be altering the sound in multiple passes.

**Comment by LWinterberg** [comment]
Such a feature would be essential for music production (conforming audio to beats and measures), so it's definitely on the roadmap. However, it is in the somewhat distant future as there's a lot of things which need to happen before that.
